Safety/Src/imu.cpp

```diff
@@ -1,119 +1,256 @@
+
+	Data.accel_x = (static_cast<float>(accx) / ACC_RANGE_8_FACTOR) - IMUCalibration.accel_x;
+	Data.accel_y = (static_cast<float>(accy) / ACC_RANGE_8_FACTOR) - IMUCalibration.accel_y;
+	Data.accel_z = (static_cast<float>(accz) / ACC_RANGE_8_FACTOR) - IMUCalibration.accel_z;
+
+	Data.gyro_x = (static_cast<float>(gyrx) / GYRO_RANGE_1000_FACTOR) - IMUCalibration.gyro_x;
+	Data.gyro_y = (static_cast<float>(gyry) / GYRO_RANGE_1000_FACTOR) - IMUCalibration.gyro_y;
+	Data.gyro_z = (static_cast<float>(gyry) / GYRO_RANGE_1000_FACTOR) - IMUCalibration.gyro_z;
 }
 
-/* Private methods  ---------------------------------------------------------*/
+BMX160::BMX160() {
+	HAL_I2C_Init(&hi2c1);
+	IMUInit();
+	this -> calibrate(); // Calibrate IMU and populate callibration struct for use
+}
 
-MPU6050::MPU6050(){
-  HAL_I2C_Init(&hi2c1);
-  ImuInit();
+void BMX160::IMUInit(void) {
+	if (scan()) {
+		setAllPowerModesToNormal();
+		configAcc();
+		configGyro();
+		configMag();
+	}
+	else {
+		// The device doesn't exist or isn't responding - there is some error
+		// TODO: Implement error protocol
+	}
 }
 
-void MPU6050::ImuInit (void){
-	//Use this for IMU set up
-	uint8_t imu_check = 0, data_check = 0;
-	HAL_StatusTypeDef wake = HAL_OK;
+void BMX160::setAllPowerModesToNormal(){
+	// Set gyro to normal mode
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, CMD_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) GYRO_NORMAL_MODE_CMD, 1, HAL_MAX_DELAY);
 
+	// Set accelerometer to normal mode
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, CMD_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) ACC_NORMAL_MODE_CMD, 1, HAL_MAX_DELAY);
 
-	//Check if device is present and the IMU is the expected device
-	
-		if (HAL_I2C_Mem_Read(&hi2c1, IMU_ADDRESS, WHO_AM_I, 1, &imu_check, 1, HAL_MAX_DELAY) == HAL_OK && imu_check == kImuAddr){
+	// Set magnetometer to normal mode
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, CMD_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) MAG_NORMAL_MODE_CMD, 1, HAL_MAX_DELAY);
 
-				//Wake the sensor up
-				wake = HAL_I2C_Mem_Write(&hi2c1, IMU_ADDRESS, PWR_MGMT_1, 1, &data_check, 1, HAL_MAX_DELAY);
+}
+
+void configAcc() {
+	// Configure acceleration sampling rate as 800 Hz and every four are averaged
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, ACC_CONF_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) ACC_ODR_800_OSR4, 1, HAL_MAX_DELAY);
 
-				data_check = 0x07;
-				HAL_I2C_Mem_Write(&hi2c1, IMU_ADDRESS, SMPLRT_DIV, 1, &data_check, 1, HAL_MAX_DELAY);
+	// Configure accelerometer to have range of +- 8g
+	HAL_I2C_Mem_Write(&hi2c1, BMX160_I2C_ADDR | BMX160_WRITE_BIT, ACC_RANGE_REG, I2C_MEMADD_SIZE_8BIT, (uint8_t*) ACC_RANGE_8G, 1, HAL_MAX_DELAY);
+}
 
-				//Configure gyro to have range of +/- 250 dps
-				data_check = 0x00;
-				HAL_I2C_Mem_Write(&hi2c1, IMU_ADDRESS, GYRO_CONFIG, 1, &data_check, 1, HAL_MAX_DELAY);
+void configGyro() {
+	// Configure gyro sampling rate as 800 Hz and every four samples are averaged
```

Comment:
do we want to average here or have sf handle this

Reply:
SF is going to run at 50Hz because the processor is too slow for 200Hz. Is there anything that needs to change here?

Reply:
No change required. Readings seem to be good as is.